trunner/harness/base.py

```diff
@@ -0,0 +1,149 @@
+import time
+from abc import ABC, abstractmethod
+from typing import Callable, Optional
+
+from trunner.text import bold
+from trunner.host import Host
+from trunner.dut import Dut
+from trunner.types import TestResult
+
+
+class HarnessError(Exception):
+    """Base class for errors thrown in harnesses."""
+
+    pass
+
+
+class ProcessError(HarnessError):
+    name: str = "PROCESS"
+
+    def __init__(self, msg: str = "", output: Optional[str] = None):
+        self.msg = msg
+        self.output = output
+
+    def __str__(self):
+        # TODO format github actions output
+        err = [bold(f"{self.name} ERROR:"), self.msg]
+
+        if self.output is not None:
+            err.extend([bold("OUTPUT:"), self.output])
+
+        err.append("")
+        return "\n".join(err)
+
+
+class FlashError(ProcessError):
+    name: str = "FLASH"
+
+
+class Rebooter:
+    """Class that provides all necessary methods needed for rebooting target device."""
+
+    def __init__(self, host: Host, dut: Dut):
+        self.host = host
+        self.dut = dut
+
+    def _reboot_soft(self):
+        self.host.set_reset(0)
+        time.sleep(0.05)
+        self.dut.clear_buffer()
+        self.host.set_reset(1)
+
+    def _reboot_hard(self):
+        self.host.set_power(False)
+        time.sleep(0.5)
+        self.dut.clear_buffer()
+        self.host.set_power(True)
+        time.sleep(0.5)
+
+    def _reboot_dut_gpio(self, hard):
+        if hard:
+            self._reboot_hard()
+        else:
+            self._reboot_soft()
+
+    def _reboot_dut_text(self, hard):
+        pass
+
+    def _set_flash_mode(self, flash):
+        pass
+
+    def __call__(self, flash=False, hard=False):
+        """Sets flash mode and perform hard or soft reboot based on `hard` flag."""
+
+        self._set_flash_mode(flash)
+
+        if self.host.has_gpio():
+            self._reboot_dut_gpio(hard=hard)
+        else:
+            # Perform rebooting with user interaction
+            self._reboot_dut_text(hard=hard)
+
+
+class HarnessBase(ABC):
+    """Base class for harnesses functors that are tied together."""
+
+    @abstractmethod
+    def __call__(self):
+        """Implements the harness logic. Every class that inherites must to implement this method."""
+
+
+class TermHarness(HarnessBase):
```

Comment:
it took me a long while to undestand term stands for terminal could you rename it to TerminalHarness?

Reply:
done